test/unit/math/test_ad_matvar.hpp

```diff
@@ -0,0 +1,724 @@
+    expect_near_rel(
+        message + std::string(" values at i = ") + std::to_string(i),
+        x[i].val(), y[i].val(), tols.gradient_val_);
+    expect_near_rel(
+        message + std::string(" adjoints at i = ") + std::to_string(i),
+        x[i].adj(), y[i].adj(), tols.gradient_grad_varmat_matvar_);
+  }
+}
+
+/**
+ * Overload for non-`std::vector` arguments with scalar type `var`
+ *
+ * @tparam T1 Type of first argument, will have scalar type `var`
+ * @tparam T2 Type of second argument, will have scalar type `var`
+ * @param message Debug message to print on fail
+ * @param x First argument to compare
+ * @param y Second argument to compare
+ * @param tols Tolerances for comparison
+ */
+template <typename T1, typename T2, require_all_st_var<T1, T2>* = nullptr,
+          require_all_not_std_vector_t<T1, T2>* = nullptr>
+void expect_near_rel_matvar(const std::string& message, T1&& x, T2&& y,
+                            const ad_tolerances& tols) {
+  expect_near_rel(message + std::string(" values"), x.val(), y.val(),
+                  tols.gradient_val_);
+  expect_near_rel(message + std::string(" adjoints"), x.adj(), y.adj(),
+                  tols.gradient_grad_varmat_matvar_);
+}
+
+/**
+ * Overload for arguments with arithmetic scalar type. This only checks
+ * values since the adjoints don't exist
+ *
+ * @tparam T1 Type of first argument, will have arithmetic scalar type
+ * @tparam T2 Type of second argument, will have arithmetic scalar type
+ * @param message Debug message to print on fail
+ * @param x First argument to compare
+ * @param y Second argument to compare
+ * @param tols Tolerances for comparison
+ */
+template <typename T1, typename T2,
+          require_any_st_arithmetic<T1, T2>* = nullptr>
+void expect_near_rel_matvar(const std::string& message, T1&& x, T2&& y,
+                            const ad_tolerances& tols) {
+  expect_near_rel(message + std::string(" doubles"), stan::math::value_of(x),
+                  stan::math::value_of(y), tols.gradient_val_);
+}
+
+/**
+ * This is the implementation of the overload for tuples of arguments
+ *
+ * @tparam T1 Tuple type of first argument
+ * @tparam T2 Tuple Type of second argument
+ * @param message Debug message to print on fail
+ * @param x First argument to compare
+ * @param y Second argument to compare
+ * @param tols Tolerances for comparison
+ * @param i Index sequence for accessing tuple
+ */
+template <typename... T1, typename... T2, std::size_t... I>
+void expect_near_rel_matvar(const std::string& message,
+                            const std::tuple<T1...>& x,
+                            const std::tuple<T2...>& y,
+                            const ad_tolerances& tols,
+                            std::index_sequence<I...> i) {
+  std::vector<int> A(
+      {(expect_near_rel_matvar(
+            message + std::string(", argument ") + std::to_string(I + 1),
+            std::get<I>(x), std::get<I>(y), tols),
+        0)...});
```

Comment:
I'd use something like the below instead of storing them in a vector

https://gist.github.com/SteveBronder/410f7127f8be9dced2b1915765fd8644#file-for_each-hpp-L19

Reply:
Resolved, but I snuck and added the `index` as an argument to pass through with `for_each`. This way I don't have to worry about index sequences or anything it's just all handled by `for_each`. It creates an extra argument tho.